src/nvapi_gpu.cpp

```diff
@@ -469,13 +469,50 @@ extern "C" {
         if (pGpuArchInfo->version != NV_GPU_ARCH_INFO_VER_1 && pGpuArchInfo->version != NV_GPU_ARCH_INFO_VER_2)
             return IncompatibleStructVersion(n);
 
-        auto architectureId = adapter->GetArchitectureId();
+        NV_GPU_ARCHITECTURE_ID architectureId = {};
+
+        if (auto archOverride = env::getEnvVariable("DXVK_NVAPI_GPU_ARCH"); !archOverride.empty()) {
+#define CHECK_ARCH(arch)       \
+    if (archOverride == #arch) \
+        architectureId = NV_GPU_ARCHITECTURE_##arch;
+            CHECK_ARCH(T2X)
+            CHECK_ARCH(T3X)
+            CHECK_ARCH(T4X)
+            CHECK_ARCH(T12X)
+            CHECK_ARCH(NV40)
+            CHECK_ARCH(NV50)
+            CHECK_ARCH(G78)
+            CHECK_ARCH(G80)
+            CHECK_ARCH(G90)
+            CHECK_ARCH(GT200)
+            CHECK_ARCH(GF100)
+            CHECK_ARCH(GF110)
+            CHECK_ARCH(GK100)
+            CHECK_ARCH(GK110)
+            CHECK_ARCH(GK200)
+            CHECK_ARCH(GM000)
+            CHECK_ARCH(GM200)
+            CHECK_ARCH(GP100)
+            CHECK_ARCH(GV100)
+            CHECK_ARCH(GV110)
+            CHECK_ARCH(TU100)
+            CHECK_ARCH(GA100)
+            CHECK_ARCH(AD100)
+#undef CHECK_ARCH
+
+            if (architectureId)
+                log::info(str::format(n, ": GPU Architecture overriden to ", archOverride));
+        }
 
-        if (env::needsAmpereSpoofing(architectureId, returnAddress))
-            architectureId = NV_GPU_ARCHITECTURE_GA100;
+        if (!architectureId) {
+            architectureId = adapter->GetArchitectureId();
 
-        if (env::needsPascalSpoofing(architectureId))
-            architectureId = NV_GPU_ARCHITECTURE_GP100;
+            if (env::needsAmpereSpoofing(architectureId, returnAddress))
```

Comment:
That would be a bit impractical. If we squashed both env-based override and profile-based spoofing into one, we would need to call `adapter->GetArchitectureId()` before the second part anyway so there's no point returning `std::optional` because we already have the final result. How about `NV_GPU_ARCHITECTURE_ID env::getGpuArchIdAfterOverrides(NvapiAdapter* adapter, void* returnAddress)`?

Reply:
What I meant to say was to keep only a snippet like this in `NvAPI_GPU_GetArchInfo`:

```c++
// snip
        if (pGpuArchInfo->version != NV_GPU_ARCH_INFO_VER_1 && pGpuArchInfo->version != NV_GPU_ARCH_INFO_VER_2)
            return IncompatibleStructVersion(n);

        auto architectureId = adapter->GetArchitectureId();
        auto architectureIdOverride = env::needsArchitectureSpoofing(architectureId, returnAddress);
        if (architectureIdOverride.has_value())
            architectureId = architectureIdOverride.value();

        // Assume the implementation ID from the architecture ID. No simple way
        // to do a more fine-grained query at this time. Would need wine-nvml
        // usage.

// snip
``` 

and do everything else within `env::needsArchitectureSpoofing`. I think it is fine to always get the architecture id from the adapter before calling that function, even if some code path within `env::needsArchitectureSpoofing` will not use it. `NvAPI_GPU_GetArchInfo` is (I hope) nowhere on a hot path and we've already fetched all the Vulkan structures, so `NvapiAdapter::GetArchitectureId()` isn't expensive.

Reply:
I see. What about changing the log line for `DXVK_NVAPI_ALLOW_OTHER_DRIVERS` into something  like this:

`log::info(str::format(allowOtherDriversEnvName, " is set, using Pascal architecture for GPU with non-NVIDIA proprietary driver"));`

and later on log something like:

`log::info(str::format(gpuArchEnvName, " is set to '", arch , "', spoofing this GPU architecture instead of detected architecture"));`

to (hopefully) avoid to much confusion.